fix: fix #181 - use measureInWindow of the ref instead of NativeModules

arpitdalal opened the pull request:
`NativeModules` is not working in the new architecture, so using the `measureInWindow` method on the ref as suggested by the react-native documentation [here](https://reactnative.dev/blog/2024/10/23/the-new-architecture-is-here#uselayouteffect) and [here](https://reactnative.dev/docs/legacy/direct-manipulation#measureinwindowcallback).

mlazari commented:
@arpitdalal Is this a breaking change for apps that still didn't enable new architecture or it works for those too?

arpitdalal commented:
Hey, just to clarify @arpitdalal-hp and I are the same person with different profiles. Sorry for the confusion but I didn't notice I created the PR with a different profile than the issue.
I pushed a [commit](https://github.com/SteffeyDev/react-native-popover-view/pull/182/commits/2669d097d9cc9eb19e4898b6d6c1b8cb672e46a3) to address your comment, initially I removed the usage of `NativeModules`, but to support older react-native versions, I've added it back with a check to use it if `measureInWindow` doesn't exist on `ref`.

smfunder commented:
hey @arpitdalal I just created a bug here with a patch file, but didn't find your PR. But we are fixing the exact same issue.

Here is my bug report: https://github.com/SteffeyDev/react-native-popover-view/issues/183

AdamGerthel commented:
Has anyone tested this to confirm that it's working for both old and new arch, and both Android and iOS?

mlazari commented:
@AdamGerthel I tested this in a project today with both Android and iOS, on both old and new architecture, and it works on both for me, but on Android it does some weird jumping when I open it first time, and it's positioned a little to high compared to where it is with old arch without this fix (same behaviour on both old and new arch with this fix). Tested with this patch (I generated the dist folder from this PR to use in the patch): 
[react-native-popover-view+5.1.9.patch](https://github.com/user-attachments/files/18096077/react-native-popover-view%2B5.1.9.patch)

AdamGerthel commented:
@mlazari I'm actually testing it right now. The position is incorrect on Android on old arch (have not tested new arch yet) as you mentioned. See below:

### Before applying the PR

https://github.com/user-attachments/assets/ec488141-72d5-46b8-a922-f6df170ec8df

### After applying the PR

https://github.com/user-attachments/assets/9111f0e0-df8e-43bd-ae00-c1e2cde58760

I'm not sure why the original behavior has been changed because I'm not at all acquainted with this code myself.

@arpitdalal I can't merge in the current state since it introduces a bug in old arch.

arpitdalal commented:
Hey @AdamGerthel @mlazari,
Thank you both for looking into this PR and issue. I appreciate you adding the information to the contributing guide and testing my changes @AdamGerthel. I couldn't test it on my local machine because my firewall does not expose expo ports and the Expo Go iOS app doesn't support expo SDK 44 anymore. But after viewing the issues you shared @AdamGerthel, I believe I should've used `measure` instead of `measureInWindow` on the `ref` as `measureInWindow` accounts for the height of the status bar on Android and reduces the `Y` number it returns back. I've made this commit ca1f9ab9b69a223db958954eb90f36bbe2659e61 to address it, can I please ask you to test it on your end because as I said earlier, I cannot test it on my local machine nor can I test it using Expo Go app.

mlazari commented:
@AdamGerthel I just opened a PR that updates the dependency versions in the test app: https://github.com/SteffeyDev/react-native-popover-view-test-app/pull/27

mlazari commented:
@AdamGerthel Actually I remembered what that issue that I see on Android is related to: https://github.com/SteffeyDev/react-native-popover-view/issues/156#issuecomment-1853813272

Looks like the StatusBar is translucent by default on your Android emulator / phone and not translucent in my emulator.

mlazari commented:
@arpitdalal I tested your latest change and now the positioning is completely wrong and in some cases (e.g. tapping the "Top" and "Left" buttons in the test app) the UI freezes:

https://github.com/user-attachments/assets/be8b811f-a8c1-4e4c-a53b-19bd2e3b0ecd

https://github.com/mlazari/react-native-popover-view-test-app/commit/8647d008932398778ac6ed412055f5afeaeb4292#diff-6f9b056b9f5abcf169a2a225ce7fcba763d303776b16ed191549bcfa0091edb6

AdamGerthel commented:
FYI I've pushed some updates to [react-native-popover-view-test-app](https://github.com/SteffeyDev/react-native-popover-view-test-app) which should make it easier to test things.

mlazari commented:
@arpitdalal This PR without the latest commit (in other words the changes at commit https://github.com/SteffeyDev/react-native-popover-view/pull/182/commits/2669d097d9cc9eb19e4898b6d6c1b8cb672e46a3) seem to work fine when the status bar is not translucent, as it is in the latest version of the test app:

https://github.com/user-attachments/assets/5cbcaae9-624b-4d51-9c46-40a4d40562d2

When it's translucent though, it's shown too high (the difference seems to be the height of the status bar, so we probably need to add the height of the status bar in that case):

https://github.com/user-attachments/assets/efb6951a-f608-4db8-a5a3-ffbb1392735c

Tested only old arch so far.

mlazari commented:
It is the same behaviour with the new architecture enabled:

Non-translucent status bar (looks good):

https://github.com/user-attachments/assets/e153d3c9-f987-4c1c-8c20-06b2d8f2d375

Translucent status bar (too high on Android):

https://github.com/user-attachments/assets/b2cfa3c7-0c65-4d9c-86b8-1e608b5e0ad9

arpitdalal commented:
Was this translucent header an issue before my PR too?

mlazari commented:
> Was this translucent header an issue before my PR too?

When `statusBarTranslucent` has the same value as the translucent status of the StatusBar, it was positioned correctly before this PR.

mlazari commented:
@arpitdalal This is how it looks without the changes in this PR (works on old arch only):

Non-translucent status bar (looks good):

https://github.com/user-attachments/assets/1b06c34b-08ff-4ba8-9edd-451ee156a004

Translucent status bar (looks good):

https://github.com/user-attachments/assets/2e12c274-f249-42d1-b894-c254c0c67a87

mlazari commented:
I logged the values returned by `ref.current.measureInWindow(...)` vs `NativeModules.UIManager.measure(...)` in the emulator I am using when clicking on the "Right" button in the test app, right after opening the app, without scrolling anything, with status bar translucent and non-translucent:
```
***> StatusBar.currentHeight 49.45454406738281

-----------------------------------------

Status Bar translucent:

-----------------------------------------

***> ref.current.measureInWindow:  {
  "x": 51.272727966308594,
  "y": 186.18182373046875,
  "width": 62.54545593261719,
  "height": 49.45454406738281
}

***> NativeModules.UIManager.measure:  {
  "x": 51.272727966308594,
  "y": 235.63636779785156,
  "width": 62.54545593261719,
  "height": 49.45454406738281
}

-----------------------------------------

Status Bar non-translucent:

-----------------------------------------

***> ref.current.measureInWindow:  {
  "x": 51.272727966308594,
  "y": 235.63636779785156,
  "width": 62.54545593261719,
  "height": 49.45454406738281
}

***> NativeModules.UIManager.measure:  {
  "x": 51.272727966308594,
  "y": 235.63636779785156,
  "width": 62.54545593261719,
  "height": 49.45454406738281
}

-----------------------------------------
```

Looks like with the old `NativeModules.UIManager.measure` the `y` value stays the same no matter if the status bar is translucent or not, but with the new `ref.current.measureInWindow` the value of StatusBar.currentHeight is substracted from `y` when the status bar is translucent.

I think a potential solution would be to ignore the value of the `statusBarTranslucent` prop when `ref.current.measureInWindow` is used and always consider it false.

mlazari commented:
This change works for me on both old and new architecture, on both iOS and Android, with both status bar translucent and non-translucent: https://github.com/SteffeyDev/react-native-popover-view/pull/185/files

arpitdalal commented:
Awesome, thank you @mlazari for looking into this and proposing a fix 👍

AdamGerthel commented:
Closing in favor of https://github.com/SteffeyDev/react-native-popover-view/pull/185
